[SPARK-8103][core] DAGScheduler should not submit multiple concurrent attempts for a stage

squito opened the pull request:
https://issues.apache.org/jira/browse/SPARK-8103

cc @kayousterhout (thanks for the extra test case)

kayousterhout committed:
Unit tests for concurrent stages issue

squito committed:
wip

squito committed:
wip on getting a better test case ...

squito committed:
better fix and simpler test case

squito committed:
ignore for jenkins

squito committed:
unit test just to make sure we fail fast on concurrent attempts

squito committed:
update test since listenerBus.waitUntilEmpty now throws an exception instead of returning a boolean

squito committed:
get rid of more random test case since kays tests are clearer

squito committed:
style

squito commented:
(separate comment just to avoid putting this in the pr description for eventual merging)

@kayousterhout I agree that your test is much clearer than running jobs with various pauses.  I had tried to come up with something clearer but your version is much better.

I am a little torn, however -- this is certainly better for understanding this narrow case, but I feel it also gives a false sense of confidence b/c you only cover a narrow set of cases.  Eg., your new case covers the extra failure happening after the reduce stage is retried, but not if the failure happens before the reduce stage is retried, or after the reduce stage retry has completely finished.  To address that, I've been running a a variety of tests like this https://github.com/squito/spark/blob/combined_multi_attempt_fixes/core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerFailureRecoverySuite.scala#L39
with some randomization in the parameters.  FWIW, with these changes and https://github.com/apache/spark/pull/6648, those tests pass consistently for me.  Of course, that is still getting limited coverage, and I'd like to expand it even more.  I'd like to find a better home for running tests like that (not just running them periodically on my laptop), but I guess that is separate issue.  So I think this can be merged as is, w/ the tests you wrote.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34655 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34655/consoleFull) for   PR 6750 at commit [`b6bc248`](https://github.com/apache/spark/commit/b6bc2489e8115fdb9b75dd6ab38082a08c546bae).

SparkQA commented:
[Test build #34655 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34655/console) for   PR 6750 at commit [`b6bc248`](https://github.com/apache/spark/commit/b6bc2489e8115fdb9b75dd6ab38082a08c546bae).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito committed:
debugging printlns

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34665 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34665/consoleFull) for   PR 6750 at commit [`ecb4e7d`](https://github.com/apache/spark/commit/ecb4e7db817af6199642e96a9f6465b8f4b695d9).

SparkQA commented:
[Test build #34665 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34665/console) for   PR 6750 at commit [`ecb4e7d`](https://github.com/apache/spark/commit/ecb4e7db817af6199642e96a9f6465b8f4b695d9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito committed:
more debug printlns

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34689 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34689/consoleFull) for   PR 6750 at commit [`9601b47`](https://github.com/apache/spark/commit/9601b47b1fd84a0e371173d4aed4632b24f14580).

SparkQA commented:
[Test build #34689 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34689/console) for   PR 6750 at commit [`9601b47`](https://github.com/apache/spark/commit/9601b47b1fd84a0e371173d4aed4632b24f14580).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito committed:
more printlns ...

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34692 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34692/consoleFull) for   PR 6750 at commit [`89a59b6`](https://github.com/apache/spark/commit/89a59b6a6ece1ec3d0ed3b3855f4daca094932c0).

SparkQA commented:
[Test build #34692 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34692/console) for   PR 6750 at commit [`89a59b6`](https://github.com/apache/spark/commit/89a59b6a6ece1ec3d0ed3b3855f4daca094932c0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito committed:
Merge branch 'master' into SPARK-8103

squito committed:
finally found the issue ... clean up debug stuff

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito committed:
more cleanup of debug garbage

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34695 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34695/consoleFull) for   PR 6750 at commit [`46bc26a`](https://github.com/apache/spark/commit/46bc26aadfe1dcd5cbf37b8d3754bbe019d9664e).

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #34695 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34695/console) for   PR 6750 at commit [`46bc26a`](https://github.com/apache/spark/commit/46bc26aadfe1dcd5cbf37b8d3754bbe019d9664e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito committed:
Merge branch 'master' into SPARK-8103

Conflicts:
	core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

SparkQA commented:
[Test build #36042 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36042/consoleFull) for   PR 6750 at commit [`d8eb202`](https://github.com/apache/spark/commit/d8eb202eb3e91560357c40c205b13400559b2315).

SparkQA commented:
[Test build #36042 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36042/console) for   PR 6750 at commit [`d8eb202`](https://github.com/apache/spark/commit/d8eb202eb3e91560357c40c205b13400559b2315).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36118 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36118/consoleFull) for   PR 6750 at commit [`d8eb202`](https://github.com/apache/spark/commit/d8eb202eb3e91560357c40c205b13400559b2315).

SparkQA commented:
[Test build #36118 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36118/console) for   PR 6750 at commit [`d8eb202`](https://github.com/apache/spark/commit/d8eb202eb3e91560357c40c205b13400559b2315).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

squito commented:
ping @kayousterhout @markhamstra @mateiz

markhamstra commented on core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:
```diff
@@ -163,6 +163,12 @@ private[spark] class TaskSchedulerImpl(
+        if (taskSets.size > 1) {
+          throw new SparkIllegalStateException("more than one active taskSet for stage " + stage)
+        }
+      }
```
@kayousterhout How much of a concern should the extra overhead be here?  Just wondering whether this (let's hope rare) condition might better be handled only in a non-production environment and behind an `if(debug)` kind of flag.

vanzin replied:
Perhaps the code could just look for an existing task set that matches the stage ID of the task set being added? That should be a little better than the filter / groupBy. Something like:

    activeTaskSets.exists { case (_, ts) => ts.stageId == taskSet.stageId }

markhamstra commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1002,6 +1001,7 @@ class DAGScheduler(
     val stageId = task.stageId
     val taskType = Utils.getFormattedClassName(task)
 
+    // REVIEWERS: does this need special handling for multiple completions of the same task?
```
Doesn't look like it to me.  The OutputCommitCoordinator should deny any completion from other than the currently authorized Stage, and should only authorize one Task per partition to commit per Stage.

squito replied:
great, thanks mark.  I'll leave the comment in for now just to get any other reviewers to take a look too.  Think this deserve some special handling in the test cases?  (and if so, any suggestions / pointers on what to do?)  I'm not very familiar w/ the OutputCommitCoordinator so guidance is appreciated.

markhamstra commented:
Other than the question of whether doing more while holding the TaskSchedulerImpl lock is a concern, this LGTM.  I'll come back later to take a closer look at the new tets in DAGSchedulerSuite.

vanzin commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -547,6 +554,136 @@ class DAGSchedulerSuite
+      createFakeTaskInfo(),
+      null))
+
+    // Another ResubmitFailedStages event should not result result in another attempt for the map
```
nit: result result

vanzin commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -547,6 +554,136 @@ class DAGSchedulerSuite
+    sc.listenerBus.waitUntilEmpty(WAIT_TIMEOUT_MILLIS)
+    assert(countSubmittedMapStageAttempts() === 2)
+
+    // NOTE: the actual ResubmitFailedStages may get called at any time during this, shouldn't
```
Should this comment be right before L612?

vanzin commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -547,6 +554,136 @@ class DAGSchedulerSuite
     assert(sparkListener.failedStages.size == 1)
   }
 
+  /** This tests the case where another FetchFailed comes in while the map stage is getting
```
nit: we generally use javadoc-style (even though this is not a method declaration):

    /**
     * Comment.
     */

vanzin commented on core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:
```diff
@@ -128,4 +128,20 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with L
+    val attempt1 = new TaskSet(Array(new FakeTask(0)), 0, 0, 0, null)
+    val attempt2 = new TaskSet(Array(new FakeTask(0)), 0, 1, 0, null)
+    taskScheduler.submitTasks(attempt1)
+    intercept[SparkIllegalStateException] { taskScheduler.submitTasks(attempt2)}
```
nit: missing space before `}`

vanzin commented:
Looks sane to me, although I'm not really familiar with all the surrounding code.

squito committed:
reviewer feedback

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:
```diff
@@ -163,6 +163,15 @@ private[spark] class TaskSchedulerImpl(
+      }
+      if (conflictingTaskSet) {
+        throw new SparkIllegalStateException(s"more than one active taskSet for stage $stage")
+      }
```
restoring the comments from the old diff b/c they are still relevant:

from mark:
> @kayousterhout How much of a concern should the extra overhead be here? Just wondering whether this (let's hope rare) condition might better be handled only in a non-production environment and behind an if(debug) kind of flag.

from marcelo:
>Perhaps the code could just look for an existing task set that matches the stage ID of the task set being added? That should be a little better than the filter / groupBy.

good point, there isn't any need to do the `groupBy`, so I've made it simpler.

I'd really rather leave the check in place.  In fact I think this fail-fast behavior is especially important in a *production* environment -- that's much better than an infinite loop of failures.

markhamstra replied:
Yes, I totally agree that it is valuable to catch the failure in production.  I'm only suggesting that at some point the check becomes a big enough performance hit that it makes sense to compromise on the fail-fast desiderata in order to maintain production performance while trying to ensure in development that the failure can never occur.  I doubt that this check is that costly, but my expectation is that Kay has a better sense of how much more we can afford to do within this synchronized block.

kayousterhout replied:
What if you change activeTaskSets to be keyed on the stageId? That seems pretty easy to do and would make this check O(1) rather than O(N)

SparkQA commented:
[Test build #36300 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36300/consoleFull) for   PR 6750 at commit [`ada7726`](https://github.com/apache/spark/commit/ada7726d4599875355114e59a7d5444f05df4685).

kayousterhout commented:
Looking at this is on my to-do list...I'll take a look this afternoon or tomorrow!

SparkQA commented:
[Test build #36300 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36300/console) for   PR 6750 at commit [`ada7726`](https://github.com/apache/spark/commit/ada7726d4599875355114e59a7d5444f05df4685).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36319 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36319/consoleFull) for   PR 6750 at commit [`ada7726`](https://github.com/apache/spark/commit/ada7726d4599875355114e59a7d5444f05df4685).

SparkQA commented:
[Test build #36319 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36319/console) for   PR 6750 at commit [`ada7726`](https://github.com/apache/spark/commit/ada7726d4599875355114e59a7d5444f05df4685).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

kayousterhout commented on core/src/main/scala/org/apache/spark/SparkException.scala:
```diff
@@ -30,3 +30,13 @@ class SparkException(message: String, cause: Throwable)
+ * Exception indicating an error internal to Spark -- it is in an inconsistent state, not due
+ * to any error by the user
+ */
+class SparkIllegalStateException(message: String, cause: Throwable)
```
Why not use the existing IllegalStateException (the non-spark specific one)? That's what done elsewhere in Spark.

squito replied:
I was hoping to create an exception which would make it clear that the user is totally free from blame.  I find IllegalStateExceptions can be confusing -- did this condition result because the user violated some precondition, misused the api etc?  Or is it a spark bug?

I felt this would be more clear, but its not critical.

kayousterhout replied:
I'd prefer to stick with the existing strategy, which is to use IllegalStateExceptions and introduce this new exception in a separate PR (my understanding is that we do currently use IllegalStateExceptions only for things that aren't the user's fault; if not, agree with your sentiment that it would be nice to clean that up, but not as part of this PR).

kayousterhout commented on core/src/main/scala/org/apache/spark/TaskContextImpl.scala:
```diff
@@ -30,6 +30,7 @@ private[spark] class TaskContextImpl(
     override val attemptNumber: Int,
     override val taskMemoryManager: TaskMemoryManager,
     val runningLocally: Boolean = false,
+    val stageAttemptId: Int = 0,  // for testing
```
Where is this used?

squito replied:
oops, good catch.  this was used in earlier versions of tests before i switched to your test.  thanks, I've updated it now.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito committed:
remove extra stageAttemptId

SparkQA commented:
[Test build #36434 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36434/consoleFull) for   PR 6750 at commit [`6542b42`](https://github.com/apache/spark/commit/6542b4297e31d02cd1d0e808574f222fa9c2fbc1).

SparkQA commented:
[Test build #36434 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36434/console) for   PR 6750 at commit [`6542b42`](https://github.com/apache/spark/commit/6542b4297e31d02cd1d0e808574f222fa9c2fbc1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36465 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36465/consoleFull) for   PR 6750 at commit [`6542b42`](https://github.com/apache/spark/commit/6542b4297e31d02cd1d0e808574f222fa9c2fbc1).

SparkQA commented:
[Test build #36465 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36465/console) for   PR 6750 at commit [`6542b42`](https://github.com/apache/spark/commit/6542b4297e31d02cd1d0e808574f222fa9c2fbc1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparkIllegalStateException(message: String, cause: Throwable)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1126,39 +1127,47 @@ class DAGScheduler(
       case FetchFailed(bmAddress, shuffleId, mapId, reduceId, failureMessage) =>
         val failedStage = stageIdToStage(task.stageId)
         val mapStage = shuffleToMapStage(shuffleId)
+        if (failedStage.attemptId - 1 > task.stageAttemptId) {
```
What's the intuition here  (I was expecting "if(failedStage.attemptId != task.stageAttemptId)")?

squito replied:
stage attempt ids get incremented when a task set is submitted (see [here](https://github.com/apache/spark/blob/1cb2629f1aa466f92246828c562ea6f35c89ab87/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L941)).  So its normal for the attempt id of tasks to be one behind `stage.attemptId`.  This magically still lets all spark listener events have the right id, b/c [we store `stage.latestInfo` before the attemptId gets incremented](https://github.com/apache/spark/blob/1cb2629f1aa466f92246828c562ea6f35c89ab87/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L875).

Honestly that logic is pretty weird to me -- i could change that instead, but I thought maybe better to not mess with that now.  I suppose the test could be `failedStage.attemptId - 1 != task.stageAttemptId`.  Maybe it would actually make the most sense to change to `failedStage.latestInfo.attemptId != task.stageAttemptId`, with a comment explaining why `failedStage.attemptId` can't be used (and maybe add a comment to `Stage.attemptId` itself)?

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -547,6 +554,138 @@ class DAGSchedulerSuite
+
+  }
+
+  /** This tests the case where a late FetchFailed comes in after the map stage has finished getting
```
change to javadoc style

kayousterhout commented:
Finally got a chance to finish looking at this; looks mostly good with a few small comments!

squito committed:
faster check for conflicting task sets

squito committed:
get rid of SparkIllegalStateException

squito committed:
style

squito committed:
remove comment for reviewers

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36697 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36697/consoleFull) for   PR 6750 at commit [`a5f7c8c`](https://github.com/apache/spark/commit/a5f7c8c0f0775529f381afa498d0a934eb0e0ff5).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36698 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36698/consoleFull) for   PR 6750 at commit [`19685bb`](https://github.com/apache/spark/commit/19685bba6d4615b9b725032a070249255caa0ff8).

squito committed:
switch to using latestInfo.attemptId, and add comments

squito commented:
@kayousterhout I think I addressed your comments.  I had some questions about the best way to make the attemptId condition cleaner -- I went ahead with one approach here https://github.com/squito/spark/commit/19685bba6d4615b9b725032a070249255caa0ff8
but I'm open to other suggestions.

Also on making the check for multiple active task sets faster -- there are a bunch of places that `activeTaskSets` is used with the existing structure.  It just makes the code more complicated to change all those usages.  So instead I added another variable for this purpose.  But I'm happy to change that to just be `activeTaskSets` if you prefer.  https://github.com/squito/spark/commit/b2faef52a5a2b49e8196a2f79f1f0c5349afd63e

SparkQA commented:
[Test build #36697 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36697/console) for   PR 6750 at commit [`a5f7c8c`](https://github.com/apache/spark/commit/a5f7c8c0f0775529f381afa498d0a934eb0e0ff5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #36698 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36698/console) for   PR 6750 at commit [`19685bb`](https://github.com/apache/spark/commit/19685bba6d4615b9b725032a070249255caa0ff8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

kayousterhout committed:
Index active task sets by stage Id rather than by task set id

kayousterhout commented:
Just checking in about this -- did you see my suggestion for using activeTaskSets rather than using a new data structure here https://github.com/squito/spark/pull/2?

Also it looks like this has fallen out of date.

squito commented:
oh, thanks for mentioning that, sorry I didn't see your suggestion!  I'll take a look later today and bring this up to date also, thanks.

squito committed:
Merge pull request #2 from kayousterhout/imran_SPARK-8103

Index active task sets by stage Id rather than by task set id

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #37230 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37230/consoleFull) for   PR 6750 at commit [`f025154`](https://github.com/apache/spark/commit/f025154199a17165206c667bdce6b5e48917cac2).

squito committed:
Revert "Index active task sets by stage Id rather than by task set id"

This reverts commit baf46e19829f9693ceb1a54457b4e1c3602ba560.

squito committed:
Merge branch 'master' into SPARK-8103

Conflicts:
	core/src/main/scala/org/apache/spark/scheduler/Stage.scala

squito committed:
fix merge

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37235 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37235/consoleFull) for   PR 6750 at commit [`906d626`](https://github.com/apache/spark/commit/906d6266efaa899b3fba949428018bd75157731a).

SparkQA commented:
**[Test build #37230 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37230/console)**     for PR 6750 at commit [`f025154`](https://github.com/apache/spark/commit/f025154199a17165206c667bdce6b5e48917cac2)     after a configured wait of `175m`.

AmplabJenkins commented:
Build finished. Test FAILed.

SparkQA commented:
[Test build #37235 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37235/console) for   PR 6750 at commit [`906d626`](https://github.com/apache/spark/commit/906d6266efaa899b3fba949428018bd75157731a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Least(children: Seq[Expression]) extends Expression `
  * `case class Greatest(children: Seq[Expression]) extends Expression `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito committed:
Merge branch 'master' into SPARK-8103

SparkQA commented:
[Test build #37252 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37252/consoleFull) for   PR 6750 at commit [`a21c8b5`](https://github.com/apache/spark/commit/a21c8b543d1ba6aec3b3dd869f294f8220995827).

SparkQA commented:
[Test build #37252 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37252/console) for   PR 6750 at commit [`a21c8b5`](https://github.com/apache/spark/commit/a21c8b543d1ba6aec3b3dd869f294f8220995827).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito committed:
get rid of activeTaskSets

SparkQA commented:
[Test build #37395 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37395/consoleFull) for   PR 6750 at commit [`c04707e`](https://github.com/apache/spark/commit/c04707ee9e177a6150eb6df250be427faec6f137).

squito committed:
add tests to make sure that TaskSchedulerImpl schedules correctly with zombie attempts

squito commented:
@kayousterhout I made the changes to get rid of the extra data structure in `TaskSchedulerImpl`, see what you think.  Also I added some test cases to `TaskSchedulerImplSuite` that cover the cases I was mentioned that I think would break with your PR.  (those tests would pass before this change in any case, but I think they're good to have.)

squito committed:
style

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:
```diff
@@ -319,26 +337,24 @@ private[spark] class TaskSchedulerImpl(
+              taskResultGetter.enqueueFailedTask(taskSet, tid, state, serializedData)
             }
           case None =>
             logError(
```
there is a small change in logic here.  Before, we'd hit this case only if `taskIdToTaskSetId` didn't contain `tid`.  If `activeTaskSets` doesn't contain `taskSetId`, we'd do nothing (foreach on None).  But now, we hit the error case if either `taskIdToTaskSetId` doesn't contain `tid`, or if `taskSetsByStageIdAndAttempt` doesn't contain the stage & attempt.

This wasn't exactly intentional, it just fell out from an attempt at refactoring a helper method -- but it does seems to more accurately correspond to what is in the error msg.  Still I'm not certain about it.  Perhaps if we nobody else has high confidence in this change I should just go back to the old behavior.

kayousterhout replied:
This is fine!

squito committed:
rename

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37396 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37396/consoleFull) for   PR 6750 at commit [`4470fa1`](https://github.com/apache/spark/commit/4470fa1aa10c902728a44da8ee06bbfca32b5108).

SparkQA commented:
[Test build #37395 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37395/console) for   PR 6750 at commit [`c04707e`](https://github.com/apache/spark/commit/c04707ee9e177a6150eb6df250be427faec6f137).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37396 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37396/console) for   PR 6750 at commit [`4470fa1`](https://github.com/apache/spark/commit/4470fa1aa10c902728a44da8ee06bbfca32b5108).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1128,38 +1128,47 @@ class DAGScheduler(
+              s"due to a fetch failure from $mapStage (${mapStage.name})")
+            markStageAsFinished(failedStage, Some(failureMessage))
+          } else {
+            logInfo(s"Ignoring fetch failure from $task as it's from $failedStage, " +
```
It seems like, even when the failed stage is no longer running, we don't fully ignore the failure, because we remove the map id from the set of output locations (on line 1164)? As a result, I'm not sure this log message makes sense.

squito replied:
yes, good point ... perhaps we should just get rid of this msg, I'm not sure if there is anything we could say here that is succint, accurate, and still useful.  I think I added this msg during some early debugging.  Do you think its worth logging "Received fetch failure from $task, but its from $failedStage which is no longer running"?  Maybe a `logDebug`?

kayousterhout replied:
Doing a logDebug seems fine if you think it will be useful! (I never turn on debug level logging just because Spark's debug logging is sooo verbose, but I can imagine that others probably do turn it on!)

kayousterhout replied:
It looks like this log statement is still lingering?

squito replied:
woops, sorry I forgot about this one!  thanks, changed to the logDebug

kayousterhout commented:
@squito I think there's a simpler fix here: I don't think TaskSchedulerImpl actually needs to know about all of the potential zombie task attempts.  Right now there are two things that happen in TaskSchedulerImpl.taskSetFinished: the task set is removed from the set of active task sets, and the schedulable is removed (for the purposes of fairness accounting).  What if we split this into two methods: one called taskSetNoLongerActive() that signals when a task set has become a zombie (so can be removed from activeTaskSets) and then rename taskSetFinished to allTasksInTaskSetFinished() or something, that removes the schedulable.  Then you could change activeTaskSets to be indexed on stage id and have a single task attempt, as in my original suggestion. I prefer that approach because it avoids tracking unnecessarily (and complex) state about all of the zombie task attempts.  Thoughts?

squito commented:
@kayousterhout I don't think that will work, but maybe I'm not seeing it.  I think the problem is, you still need some way get a handle on the zombie TaskSetManager to be able to call allTasksInTaskSetFinished().  Right now, taskSetFinished is ultimately getting a handle on that TaskSetManager by looking it up in activeTaskSets [in `statusUpdate()`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala#L328).  So if those zombie task sets aren't in activeTaskSets, it seems like you'd still need to keep track of them *somewhere* in TaskSchedulerImpl.

I feel like part of the problem is that "active" task sets is somewhat vague.  You might not expect it to contain task sets that have already failed (from a fetch failed), but still happen to have tasks running.  I guess "zombie" is vague too, but in a way that is better since you aren't tricked into thinking you know what it means.

kayousterhout commented:
Ah shoot you're right. Do you prefer your original approach of creating a new data structure to track this or the current one of making activeTaskSets more complicated?

squito commented:
I suppose I slightly prefer the current one, I think its a bit more clear what is being tracked (with so many different ids floating around, its easy to lose track of what the `id` of a taskset is).  But no strong preference.  OTOH getting to the task set manager from a task id is now a lookup in 3 maps instead of 2, dunno if that will effect performance (or if it matters here).

really I don't have any preferences ...

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1128,38 +1128,47 @@ class DAGScheduler(
-        }
+        if (failedStage.latestInfo.attemptId != task.stageAttemptId) {
+          logInfo(s"Ignoring fetch failure from $task as it's from $failedStage attempt" +
+            s" ${task.stageAttemptId}, which has already failed")
```
Can you change this to s" ${task.stageAttemptId}, and there is a more recent attempt for that stage (attempt ID ${failedStage.latestInfo.attemptId}) running"? It's a little misleading as-is, because we actually do still do something with the fetch failure if the stage has already been marked as failed, as long as there's not already a newer version of the stage running.

squito replied:
yes, good idea.  done

kayousterhout commented:
I had one last idea for how to simplify this: https://github.com/squito/spark/pull/4, that I made sure doesn't break any tests this time! Let me know what your thoughts are.

squito committed:
update log msg

squito committed:
Merge branch 'master' into SPARK-8103

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37558 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37558/consoleFull) for   PR 6750 at commit [`e43ac25`](https://github.com/apache/spark/commit/e43ac2589d2222b65366cbcd6a1d9ece724c7419).

SparkQA commented:
[Test build #37558 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37558/console) for   PR 6750 at commit [`e43ac25`](https://github.com/apache/spark/commit/e43ac2589d2222b65366cbcd6a1d9ece724c7419).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute `
  * `abstract class Star extends LeafExpression with NamedExpression `
  * `case class UnresolvedAlias(child: Expression) extends UnaryExpression with NamedExpression `
  * `case class SortOrder(child: Expression, direction: SortDirection) extends UnaryExpression `
  * `trait AggregateExpression extends Expression `
  * `trait PartialAggregate extends AggregateExpression `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate `
  * `trait Generator extends Expression `
  * `case class Explode(child: Expression) extends UnaryExpression with Generator `
  * `trait NamedExpression extends Expression `
  * `abstract class Attribute extends LeafExpression with NamedExpression `
  * `case class PrettyAttribute(name: String) extends Attribute `
  * `abstract class LeafNode extends LogicalPlan `
  * `abstract class UnaryNode extends LogicalPlan `

AmplabJenkins commented:
Merged build finished. Test PASSed.

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:
```diff
@@ -363,9 +379,13 @@ private[spark] class TaskSchedulerImpl(
-          .flatMap(activeTaskSets.get)
-          .map(taskSetMgr => (id, taskSetMgr.stageId, taskSetMgr.taskSet.attempt, metrics))
+        for {
+          (stageId, stageAttemptId) <- taskIdToStageIdAndAttempt.get(id)
```
why not use the taskSetManagerForTask(id) method here?

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -547,6 +554,139 @@ class DAGSchedulerSuite
+
+    // Another ResubmitFailedStages event should not result in another attempt for the map
+    // stage being run concurrently.
+    // NOTE: the actual ResubmitFailedStages may get called at any time during this, shouldn't
```
add "but it" before "shouldn't"

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -547,6 +554,139 @@ class DAGSchedulerSuite
+      createFakeTaskInfo(),
+      null))
+
+    // Trigger resubmission of the failed map stage and finish the re-started map task.
```
I think this comment is wrong, and should say something like "Running ResubmitFailedStages shouldn't result in any more attempts for the map stage, because the FetchFailed should have been ignored"

squito replied:
github is being weird ... I think this comment was originally on line 683 below (which I've updated), but the comment was originally duplicated.  Lemme know if I've messed up

kayousterhout replied:
Oh weird... yeah your understand is correct!

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/TaskSchedulerImplSuite.scala:
```diff
@@ -128,4 +128,113 @@ class TaskSchedulerImplSuite extends SparkFunSuite with LocalSparkContext with L
+    val taskDescriptions3 = taskScheduler.resourceOffers(workerOffers).flatten
+    assert(10 === taskDescriptions3.length)
+
+    taskDescriptions3.foreach{ task =>
```
super nit: space before "{"

kayousterhout commented:
Ok I did a last pass over this...just some last style comments, and it sounds like you're also going to do the cleanup of making the hashmap indexed on task id store the task set manager as a value? Then this looks good.

squito committed:
simplify going from taskId to taskSetMgr

squito committed:
fix some comments, style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37655 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37655/consoleFull) for   PR 6750 at commit [`e01b7aa`](https://github.com/apache/spark/commit/e01b7aa9ae3640a23ee769f7ce779332e25db5b6).

squito commented:
thanks for the last set of comments @kayousterhout .  I've updated based on your suggestions

kayousterhout commented:
There's just one lingering erroneous log message, and then this LGTM.

kayousterhout commented:
Thanks for all the iterations on this...I'm hopeful that if we made the scheduler code a little better / easier to read with each PR, these will become less painful over time.

squito committed:
fix log msg

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/32/consoleFull) for   PR 6750 at commit [`fb3acfc`](https://github.com/apache/spark/commit/fb3acfc6b7f3f5738061dbe033de33df4d84cbd9).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37663 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37663/consoleFull) for   PR 6750 at commit [`fb3acfc`](https://github.com/apache/spark/commit/fb3acfc6b7f3f5738061dbe033de33df4d84cbd9).

SparkQA commented:
[Test build #37655 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37655/console) for   PR 6750 at commit [`e01b7aa`](https://github.com/apache/spark/commit/e01b7aa9ae3640a23ee769f7ce779332e25db5b6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #37663 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37663/console) for   PR 6750 at commit [`fb3acfc`](https://github.com/apache/spark/commit/fb3acfc6b7f3f5738061dbe033de33df4d84cbd9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/33/consoleFull) for   PR 6750 at commit [`fb3acfc`](https://github.com/apache/spark/commit/fb3acfc6b7f3f5738061dbe033de33df4d84cbd9).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37668 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37668/consoleFull) for   PR 6750 at commit [`fb3acfc`](https://github.com/apache/spark/commit/fb3acfc6b7f3f5738061dbe033de33df4d84cbd9).

SparkQA commented:
[Test build #32 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/32/console) for   PR 6750 at commit [`fb3acfc`](https://github.com/apache/spark/commit/fb3acfc6b7f3f5738061dbe033de33df4d84cbd9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class StandaloneRecoveryModeFactory(conf: SparkConf, serializer: Serializer) `
  * `class RFormula(override val uid: String)`
  * `case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute `
  * `abstract class Star extends LeafExpression with NamedExpression `
  * `case class UnresolvedAlias(child: Expression) extends UnaryExpression with NamedExpression `
  * `abstract class LeafExpression extends Expression `
  * `abstract class UnaryExpression extends Expression `
  * `abstract class BinaryExpression extends Expression `
  * `case class SortOrder(child: Expression, direction: SortDirection) extends UnaryExpression `
  * `trait AggregateExpression extends Expression `
  * `trait PartialAggregate extends AggregateExpression `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Pmod(left: Expression, right: Expression) extends BinaryArithmetic `
  * `      final class SpecificRow extends $`
  * `trait Generator extends Expression `
  * `case class Explode(child: Expression) extends UnaryExpression with Generator `
  * `trait NamedExpression extends Expression `
  * `abstract class Attribute extends LeafExpression with NamedExpression `
  * `case class PrettyAttribute(name: String) extends Attribute `
  * `case class Length(child: Expression) extends UnaryExpression with ExpectsInputTypes `
  * `case class FormatNumber(x: Expression, d: Expression)`
  * `abstract class LeafNode extends LogicalPlan `
  * `abstract class UnaryNode extends LogicalPlan `
  * `abstract class BinaryNode extends LogicalPlan `

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #33 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/33/console) for   PR 6750 at commit [`fb3acfc`](https://github.com/apache/spark/commit/fb3acfc6b7f3f5738061dbe033de33df4d84cbd9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class StandaloneRecoveryModeFactory(conf: SparkConf, serializer: Serializer) `
  * `class RFormula(override val uid: String)`
  * `case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute `
  * `abstract class Star extends LeafExpression with NamedExpression `
  * `case class UnresolvedAlias(child: Expression) extends UnaryExpression with NamedExpression `
  * `abstract class LeafExpression extends Expression `
  * `abstract class UnaryExpression extends Expression `
  * `abstract class BinaryExpression extends Expression `
  * `case class SortOrder(child: Expression, direction: SortDirection) extends UnaryExpression `
  * `trait AggregateExpression extends Expression `
  * `trait PartialAggregate extends AggregateExpression `
  * `case class Min(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Max(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Count(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Average(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Sum(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class SumDistinct(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class First(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Last(child: Expression) extends UnaryExpression with PartialAggregate `
  * `case class Pmod(left: Expression, right: Expression) extends BinaryArithmetic `
  * `      final class SpecificRow extends $`
  * `trait Generator extends Expression `
  * `case class Explode(child: Expression) extends UnaryExpression with Generator `
  * `trait NamedExpression extends Expression `
  * `abstract class Attribute extends LeafExpression with NamedExpression `
  * `case class PrettyAttribute(name: String) extends Attribute `
  * `case class Length(child: Expression) extends UnaryExpression with ExpectsInputTypes `
  * `case class FormatNumber(x: Expression, d: Expression)`
  * `abstract class LeafNode extends LogicalPlan `
  * `abstract class UnaryNode extends LogicalPlan `
  * `abstract class BinaryNode extends LogicalPlan `

SparkQA commented:
[Test build #37668 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37668/console) for   PR 6750 at commit [`fb3acfc`](https://github.com/apache/spark/commit/fb3acfc6b7f3f5738061dbe033de33df4d84cbd9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.
